Userland: Add ability to screenshot rectangular region in `shot`

abyesilyurt opened the pull request:
This PR adds a new option to shot to take screenshots of rectangular regions.
User is shown a semi-transparent overlay like below, where s/he can select a rectangle with mouse clicks.

- I had to change the way fullscreen transparent windows are drawn. 
Before that, all the windows in the background were ignored in the existance of a fullscreen window.
- ~`Bitmap::cropped` function used to return a pointer to local variable,
and the returned bitmap was invalid.
I fixed it by using `static`, but there should be a better way.~
- The taskbar is not selectable (yet), related issue #8510.


![Screenshot_20210707_012653](https://user-images.githubusercontent.com/8598852/124679670-283de280-dec5-11eb-9cf2-bb9723088fa8.png)

alimpfard commented on Userland/Libraries/LibGfx/Bitmap.cpp:
```diff
@@ -476,7 +476,7 @@ RefPtr<Gfx::Bitmap> Bitmap::scaled(float sx, float sy) const
 RefPtr<Gfx::Bitmap> Bitmap::cropped(Gfx::IntRect crop) const
 {
-    auto new_bitmap = Gfx::Bitmap::create(format(), { crop.width(), crop.height() }, 1);
+    static auto new_bitmap = Gfx::Bitmap::create(format(), { crop.width(), crop.height() }, 1);
```
This is not necessary, `new_bitmap` is a RefPtr which will be copied on return.

abyesilyurt replied:
Reverted the change.

alimpfard commented on Userland/Services/WindowServer/Compositor.cpp:
```diff
@@ -1096,7 +1097,8 @@ void Compositor::recompute_occlusions()
         m_opaque_wallpaper_rects.clear();
-    } else {
+    }
+    {
```
I'm not quite sure why this (massive) else branch is now always being run?

abyesilyurt replied:
I changed this line again, can you have a look? The idea is basically to keep drawing the background, when the fullscreen window is transparent.

alimpfard replied:
looks like it does what you're saying it does, so that's alright by me :P

gunnarbeutner commented on Userland/Libraries/LibGfx/Bitmap.cpp:
```diff
@@ -476,7 +476,7 @@ RefPtr<Gfx::Bitmap> Bitmap::scaled(float sx, float sy) const
 RefPtr<Gfx::Bitmap> Bitmap::cropped(Gfx::IntRect crop) const
 {
-    auto new_bitmap = Gfx::Bitmap::create(format(), { crop.width(), crop.height() }, 1);
+    static auto new_bitmap = Gfx::Bitmap::create(format(), { crop.width(), crop.height() }, 1);
```
That doesn't look correct. For one this would return the same bitmap even if - at a later point - a different IntRect was specified.

Also, this seems unnecessary because `new_bitmap` is a heap-allocated object that's wrapped in a smart pointer (RefPtr).

Instead you should change `auto* bitmap` to `auto bitmap` in `shot.cpp` because that would otherwise destroy the RefPtr.

abyesilyurt replied:
Thanks for the correction! 

Can you help me understand why `auto* bitmap` would destroy `RefPtr`?

gunnarbeutner replied:
`Gfx::Bitmap::create` returns a `RefPtr<Gfx::Bitmap>`. When you use `auto*` that causes C++ to try to coerce whatever object you specified to the right of `=` into a pointer type.

The `RefPtr<T>` class has this operator which is then implicitly used to coerce the `RefPtr<Gfx::Bitmap>` into a `Gfx::Bitmap*`: `ALWAYS_INLINE operator T*() { return as_ptr(); }`

However, at that point the `RefPtr<Gfx::Bitmap>` that was returned by the `create()` function is no longer necessary and is implicitly destroyed (because you're not holding on to it anywhere) - which in turn also destroys the underlying `Gfx::Bitmap` object - leaving you with a dangling pointer.

abyesilyurt replied:
Thanks for the detailed info! Everything is much clear now.

alimpfard commented on Userland/Utilities/shot.cpp:
```diff
@@ -5,28 +5,100 @@
     args_parser.add_option(output_to_clipboard, "Output to clipboard", "clipboard", 'c');
     args_parser.add_option(delay, "Seconds to wait before taking a screenshot", "delay", 'd', "seconds");
     args_parser.add_option(screen, "The index of the screen (default: -1 for all screens)", "screen", 's', "index");
+    args_parser.add_option(select_region, "Select a region to capture", "region", 'r');
```
It'd also be neat to make this scriptable, so you can say `shot --rect 10,10,40,40` or something to that effect... though that can be done in a later PR/change too.

abyesilyurt replied:
Good idea! I will look into that in a later PR. Which one makes more sense to define the rect x,y,w,h or x1,y1,x2,y2?

alimpfard replied:
I personally like x,y,w,h but we can bikeshed that when needed :P

alimpfard commented on Userland/Utilities/shot.cpp:
```diff
@@ -5,28 +5,100 @@
+    Optional<Gfx::IntPoint> m_last_point {};
+    Optional<Gfx::IntPoint> m_achor_point {};
+    Optional<Gfx::IntRect> m_region {};
+    GUI::Window* m_window;
```
Also, please initialise primitive values.
(And Optional<T> has a default ctor, no need to initialise it manually)

abyesilyurt replied:
Does it look better now?

alimpfard replied:
yup yup

alimpfard commented on Userland/Utilities/shot.cpp:
```diff
@@ -5,28 +5,100 @@
 #include <LibGfx/PNGWriter.h>
 #include <unistd.h>
 
+class SelectableLayover final : public GUI::Widget {
```
It feels rather awkward to do click-click or click-click+drag to select a region, why not just have a normal region select mechanism similar to pixelpaint? i.e. click+drag to select, and release to take a screenshot?

abyesilyurt replied:
Switched to click+drag to select, the code is also much cleaner now :)

gunnarbeutner commented:
Another thing I just noticed: Can you make it so that the selected area isn't grayed out like the rest? :)

ChrisK92 commented on Userland/Utilities/shot.cpp:
```diff
@@ -5,28 +5,100 @@
+
+    static Optional<Gfx::IntRect> build_rect(Optional<Gfx::IntPoint> const& p1, Optional<Gfx::IntPoint> const& p2)
+    {
+        if (!p1.has_value() || !p1.has_value())
```
You probably want:
` if (!p1.has_value() || !p2.has_value())`

abyesilyurt replied:
Good catch :) but not necessary anymore as @MaxWipfli pointed to `Gfx::IntRect::from_two_points`.

MaxWipfli commented on Userland/Utilities/shot.cpp:
```diff
@@ -5,28 +5,100 @@
+
+    virtual void paint_event(GUI::PaintEvent&) override
+    {
+        m_region = build_rect(m_achor_point, m_last_point);
+        GUI::Painter painter(*this);
+        painter.draw_rect(m_region.value_or(Gfx::IntRect()), Gfx::Color::Black);
```
The `build_rect` function basically already exists in `Rect<T>`, so you can do this instead of writing your own function. Also, don't instantiate a `GUI::Painter` to draw an zero-size rect.
```suggestion
        if (!m_achor_point.has_value() || !m_last_point.has_value()) {
            m_region = {};
            return;
        }
        GUI::Painter painter(*this);
        m_region = IntRect::from_two_points(*m_achor_point, *m_last_point);
```

abyesilyurt replied:
Thanks! Applied the changes.

MaxWipfli commented on Userland/Utilities/shot.cpp:
```diff
@@ -5,28 +5,100 @@
+    }
+
+    Optional<Gfx::IntPoint> m_last_point {};
+    Optional<Gfx::IntPoint> m_achor_point {};
```
I guess this should be `m_anchor_point`.

MaxWipfli commented on Userland/Utilities/shot.cpp:
```diff
@@ -35,6 +107,18 @@ int main(int argc, char** argv)
+        window->set_fullscreen(true);
+        window->show();
+        app->exec();
+        crop_region = container.region();
```
I guess if there is no `crop_region`, the user has aborted and the program should thus exit? Currently, it will then just make a screenshot of the full screen.

abyesilyurt replied:
If there is no `crop_region`, the program will just exit now.

MaxWipfli commented on Userland/Utilities/shot.cpp:
```diff
@@ -5,28 +5,100 @@
+
+    virtual void keydown_event(GUI::KeyEvent& event) override
+    {
+        if (event.key() != Key_Return)
```
I didn't test this, but I would expect pressing Escape to abort the screenshot process. I think this is not the case at the moment, right?

abyesilyurt replied:
Yes, you are right. I added Escape option to quit the program without any screenshots.

MaxWipfli reviewed (changes requested):
A couple of comments. Also, please make the selected area non-grayed out, and make it so that it's Click+Drag to select, like what PixelPaint does (as Ali and Gunnar already mentioned).

abyesilyurt commented:
> Another thing I just noticed: Can you make it so that the selected area isn't grayed out like the rest? :)

Right now, all this stuff seems to be happening in `Compositor`, which is hard to follow for me.
Is it possible to implement that without any change in the  `Compositor` logic?

MaxWipfli commented:
> > Another thing I just noticed: Can you make it so that the selected area isn't grayed out like the rest? :)
> 
> Right now, all this stuff seems to be happening in `Compositor`, which is hard to follow for me as a new contributor.
> Is it possible to implement that without any change in the `Compositor` logic?

I think you need to use `fill_rect` instead of `draw_rect`, and then use `Color::Transparent` as the color.

abyesilyurt commented:
> > > Another thing I just noticed: Can you make it so that the selected area isn't grayed out like the rest? :)
> > 
> > 
> > Right now, all this stuff seems to be happening in `Compositor`, which is hard to follow for me as a new contributor.
> > Is it possible to implement that without any change in the `Compositor` logic?
> 
> I think you need to use `fill_rect` instead of `draw_rect`, and then use `Color::Transparent` as the color. Although maybe it's required to make the full window transparent, and then fill everything but the rect with some semi-transparent color.

I gave it a try, but didn't work. I think Color::Transparent only works across the widgets within the same window, i.e. you see the bottom of the window.

I also tried to make the window transparent, in that case `draw_rect` did not have any effect. Is it possible to draw on transparent windows?

abyesilyurt commented:
Thanks for all the input everyone! 

I addressed all the comments except non-graying out the selected region. I believe it is more complicated than it sounds.

alimpfard commented:
To make it not draw the selected rect, you could perhaps try not filling the window with the bg colour, and instead manually filling the regions that are meant to be grayed out, in pseudocode:
```
paint_event()
    fill_with_transparency()
    for each rect in window_rect.shatter(selected_rect):
        fill(rect, White)
    draw_rectangle(selected_rect, Black)
```

alimpfard commented on Userland/Utilities/shot.cpp:
```diff
@@ -1,32 +1,93 @@
+class SelectableLayover final : public GUI::Widget {
+    C_OBJECT(SelectableLayover)
+public:
+    SelectableLayover(GUI::Window* window)
+        : m_window(window) {};
```
Please move the empty block to a new line, and remove the extraneous `;`
```suggestion
    SelectableLayover(GUI::Window* window)
        : m_window(window)
    {
    }
```

abyesilyurt replied:
Done

abyesilyurt commented:
> To make it not fill the selected rect, you could perhaps try not filling the window with the bg colour, and instead manually filling the regions that are meant to be grayed out, in pseudocode:
> 
> ```
> paint_event()
>     fill_with_transparency()
>     for each rect in window_rect.shatter(selected_rect):
>         fill(rect, White)
>     draw_rectangle(selected_rect, Black)
> ```

Works!
![image](https://user-images.githubusercontent.com/8598852/124814495-5c1d1480-df66-11eb-8bde-ac6587002319.png)

abyesilyurt commented on Userland/Utilities/shot.cpp:
```diff
@@ -1,32 +1,103 @@
+    {
+        if (m_region.is_empty()) {
+            GUI::Painter painter(*this);
+            painter.fill_rect(m_window->rect(), m_background_color);
```
Is it correct to repaint the background everytime this function is called?

alimpfard replied:
Eh, why not :shrug:

abyesilyurt replied:
I don't know much about the internals, I thought it might be costly to call everytime.

alimpfard commented:
that image looks a little inverted though :thinking: 
shouldn't the non-selected area be grayed out?

abyesilyurt commented:
> that image looks a little inverted though 
> shouldn't the non-selected area be grayed out?

Some people suggested selected area to be non grayed out.
I think that makes sense, such that you see what you are grabbing.

alimpfard commented:
Yeah, I expect that the selected area shouldn't be grayed out, but it looks like this image does the inverse?
> ![](https://user-images.githubusercontent.com/8598852/124814495-5c1d1480-df66-11eb-8bde-ac6587002319.png)

abyesilyurt commented:
> Yeah, I expect that the selected area shouldn't be grayed out, but it looks like this image does the inverse?
> 
> > ![](https://user-images.githubusercontent.com/8598852/124814495-5c1d1480-df66-11eb-8bde-ac6587002319.png)
> 
> Update: Apparently my eyes are tricking me here

I picked Blue over Gray for the background, otherwise the contrast was very low.

alimpfard commented:
:thinking:
maybe try white or the widget base colour (`palette().base()`)?

gunnarbeutner commented on Userland/Utilities/shot.cpp:
```diff
@@ -1,32 +1,103 @@
+public:
+    SelectableLayover(GUI::Window* window)
+        : m_window(window)
+    {
```
```suggestion
    {
        set_override_cursor(Gfx::StandardCursor::Crosshair);
```

gunnarbeutner replied:
That way you get a crosshair cursor for the selection window:

![image](https://user-images.githubusercontent.com/388571/124817787-853fa400-df6a-11eb-9ad6-62cbf25452a0.png)

abyesilyurt replied:
Cool!

abyesilyurt commented:
> maybe try white or the widget base colour (`palette().base()`)?

palette().threed_highlight() gives the best contrast overall considering the other themes.

abyesilyurt commented:
This is how it looks now.

![image](https://user-images.githubusercontent.com/8598852/124818599-87563280-df6b-11eb-85fd-add9db22fbf4.png)

alimpfard commented:
:thinking: that does look nice, maybe if you bump the opacity a little bit, it'd make that a little more visible?

alimpfard commented:
Also, I just noticed that the screenshot window doesn't have a title :P 
Please give it a title lol

abyesilyurt commented:
>  that does look nice, maybe if you bump the opacity a little bit, it'd make that a little more visible?

That is the problem. Opacity is global over window, if I increase the opacity, the transparent part is effected at the same time.

abyesilyurt commented:
> Also, I just noticed that the screenshot window doesn't have a title :P
> Please give it a title lol

:D "shot"?

alimpfard commented:
> That is the problem. Opacity is global over window, if I increase the opacity, the transparent part is effected at the same time.

Then let's not block this PR, that's a separate issue anyway.

> "shot"?

Sure :)

abyesilyurt commented:
Tests are passing, can we merge?

abyesilyurt commented:
> thinking that does look nice, maybe if you bump the opacity a little bit, it'd make that a little more visible?

I opened an issue for this, #8577.

alimpfard commented:
Let's go with it
